R/extract_diagnostics.R

```diff
@@ -0,0 +1,161 @@
+#'         all chains.
+#'   \item \code{p_divergent}: The proportion of divergent transitions across
+#'         all samples.
+#'   \item \code{p_max_treedepth}: The proportion of samples that hit the
+#'         maximum tree depth.
+#'   \item \code{p_high_rhat}: The proportion of parameters with Rhat values
+#'         greater than 1.05, indicating potential convergence issues.
+#'   \item \code{low_case_count_flag}: A flag indicating if there are low case
+#'         counts in the data. See \code{low_case_count_diagnostic()} for more
+#'         information on this diagnostic.
+#'   \item \code{epinow2_diagnostic_flag}: A combined flag that indicates if
+#'         any diagnostic thresholds are exceeded.
+#' }
+#' @export
+extract_diagnostics <- function(fit,
+                                data,
+                                job_id,
+                                task_id,
+                                disease,
+                                geo_value,
+                                model) {
+  low_case_count <- low_case_count_diagnostic(data)
+
+  epinow2_diagnostics <- rstan::get_sampler_params(fit$estimates$fit,
+    inc_warmup = FALSE
+  )
+  mean_accept_stat <- mean(
+    sapply(epinow2_diagnostics, function(x) mean(x[, "accept_stat__"]))
+  )
+  p_divergent <- mean(
+    rstan::get_divergent_iterations(fit$estimates$fit),
+    na.rm = TRUE
+  )
+  p_max_treedepth <- mean(
+    rstan::get_max_treedepth_iterations(fit$estimates$fit),
+    na.rm = TRUE
+  )
+  p_high_rhat <- mean(
+    rstan::summary(fit$estimates$fit)$summary[, "Rhat"] > 1.05,
+    na.rm = TRUE
+  )
+
+  # Combine all diagnostic flags into one flag
+  diagnostic_flag <- any(
+    mean_accept_stat < 0.1,
+    p_divergent > 0.0075,
+    p_max_treedepth > 0.05,
+    p_high_rhat > 0.0075
+  )
+  # Create individual vectors for the columns of the diagnostics data frame
+  diagnostic_names <- c(
+    "mean_accept_stat",
+    "p_divergent",
+    "p_max_treedepth",
+    "p_high_rhat",
+    "diagnostic_flag",
+    "low_case_count_flag"
+  )
+  diagnostic_values <- c(
+    mean_accept_stat,
+    p_divergent,
+    p_max_treedepth,
+    p_high_rhat,
+    diagnostic_flag,
+    low_case_count
+  )
+
+  data.frame(
+    diagnostic = diagnostic_names,
+    value = diagnostic_values,
+    job_id = job_id,
+    task_id = task_id,
+    disease = disease,
+    geo_value = geo_value,
+    model = model
+  )
+}
+
+#' Calculate low case count diagnostic flag
+#'
+#' The diagnostic flag is TRUE if either of the _last_ two weeks of the dataset
+#' have fewer than an aggregate 10 cases per week. This aggregation excludes the
+#' count from confirmed outliers, which have been set to NA in the data.
+#'
+#' This function assumes that the `df` input dataset has been
+#' "completed": that any implicit missingness has been made explicit.
+#'
+#' @param df A dataframe as returned by [read_data()]. The dataframe must
```

Comment:
Why use `\code{}` sometimes and `[]` others

Reply:
....no particular reason. I mainly use `[...]` for inlining. Does it bother you?

Reply:
No I don't mind. Sorry I meant ` probably. I think basically ` is quicker than `\code{}`